### DISCUSSION TITLE
Crazyflie Bolt with Nvidia Jetson companion computer

### DISCUSSION THREAD

**Discussion by antalpeter1** [bitcraze/discussions#1313] [Q&A]
Dear All!
In our lab, we would like to implement a nonlinear MPC for trajectrory tracking with a Crazyflie Bolt based quadcopter. However, the Bolt is not powerful enough to run the algorithm in real-time, therefore we are going to equip the drone with an Nvidia Jetson companion computer. Regarding the Bolt-Jetson communication, the following questions have come to my mind:
1) Which communication interface should we use to connect the Bolt and Jetson computers? (I2C, USB, SPI, UART, ...)
2) Which protocol should we employ? Is it possible to use CRTP while also communicating with a ground control PC via radio? Or could we somehow integrate Mavlink into the setup?

We have not yet seen any solution for connecting an on-borad companion computer to the Bolt, so we appreciate all suggestions and insights.
Regards,
Peter

**Comment by knmcguire** [comment]
Hi! That's a great combination. Let me see what I can do:

1- It really depends on if you want to save on weight. With uart you can just use a few wires for tx/rx and ground if they are not sharing the same battery. And then it should be possible to do CRTP through uart via these instructions: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/. It is not super well tested and might be a bit buggy, but that is an option.
1- But what I think if you are running the Jetson computer you'll probably need more oomphf anyway as that is a bit of a power hungry beast. In that sense, the difference between some small wires and a USB cable isn't that much difference. Than you can connect the Bolt with an USB cable with the jetson and on the jetson use the crazyflie python library to communicate with it. Do be warned that I do think that there is probably some kind of 'flight blocker' that doesn't allow the bolt to fly if it is connected to USB (as usualy on the crazyflie we use that cable for charging, but not for the bolt).

2- Yes, CRTP should be possible, but also a bit of a warning there. If you would use CRTP to communicate with the Bolt through USB at your jetson, it will not longer be possible to communicate through CRTP and a crazyradio with the bolt at that point. There can only be 1-1 communication as CRTP is not build for anything else than that. In that sense, if you'd still like to communicate with the bolt directly from an external computer for flight, then setting up a simpler UART connection would be better.  There were some solutions provided in this discussion for that when people tried to connect a bolt with a px4 computer just to send through some position data. https://github.com/orgs/bitcraze/discussions/1272

**Reply by antalpeter1** [reply]
Hi Kimberly!
Thanks for your answer. 
We definitely need the radio communication for motion capture, therefore if I get it right, CRTP is not an option between the Jetson and the Bolt. In this case, I think we should go with UART. However, we would need to both send and receive data, which seems to be more complicated than the position data streaming example you've sent. 
Do you think we should use the CPX protocol over UART? I've seen that CPX has a Python implementation at https://github.com/bitcraze/crazyflie-lib-python/tree/master/cflib/cpx. Could we use that on the Jetson?
Based on the documentation, the AI deck also uses CPX over UART. Do you think we could use those code parts to write the Bolt side of the communication?

**Reply by knmcguire** [reply]
I would think so yes! As long as you don't port it to transfer crtp it can be used for other type of packets as well. Depends on how difficult the information is that you'd like to send, because if you just want to send arrays than the example that I've shown in the other discussion would work too. But it might be handy to use structures that already exist so.

Mind that we haven't that much tried this ourselves in the same scenario so do let us know if you hit any problems with using CPX
